### PR TITLE
The call_inspector application

### PR DESCRIPTION
**Pull request by fenollp**

### CODE REVIEW

File: applications/call_inspector/src/analyzers/ci_analysis.erl [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+to_json(#ci_analysis{}=Analysis) ->
+    wh_json:from_list(
+      props:filter_undefined(
+        [{<<"originate_type">>, originate_type(Analysis)}
+         ,{<<"terminate_type">>, terminate_type(Analysis)}
+         ,{<<"failure_location">>, failure_location(Analysis)}
+         ,{<<"reason">>, reason(Analysis)}
+        ])
+     );
+to_json(_) ->
```
Comment (jamesaimonetti) [inline applications/call_inspector/src/analyzers/ci_analysis.erl]: Shouldn't this not be here and the caller crash?

File: applications/call_inspector/src/parsers/ci_chunk.erl [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+                  ,parser :: atom()
+                  ,label :: ne_binary()
+                 }).
+-type chunk() :: #ci_chunk{}.
+
+-export_type([chunk/0]).
+
+-include("../call_inspector.hrl").
+
+-define(setter(Field),
```
Comment (jamesaimonetti) [inline applications/call_inspector/src/parsers/ci_chunk.erl]: We've conventionally UPPERCASED macros to set them apart visually. Unless others disagree, I'd like to continue doing so.

File: applications/call_inspector/src/parsers/ci_chunk.erl [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+              ,dst_port = DstPort
+              ,call_id = wh_json:get_value(<<"call-id">>, JObj)
+              ,timestamp = wh_json:get_value(<<"timestamp">>, JObj)
+              ,ref_timestamp = wh_util:to_float(wh_json:get_value(<<"ref_timestamp">>, JObj))
+              ,label = wh_json:get_value(<<"label">>, JObj)
+              ,data = wh_json:get_value(<<"raw">>, JObj)
+              ,parser = wh_json:get_value(<<"parser">>, JObj)
+             }.
+
+src(#ci_chunk{src_ip = Ip, src_port = Port}) ->
```
Comment (jamesaimonetti) [inline applications/call_inspector/src/parsers/ci_chunk.erl]: Missing specs here and a bunch further down.

File: applications/call_inspector/src/parsers/ci_chunk.erl [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+              ,dst_port = DstPort
+              ,call_id = wh_json:get_value(<<"call-id">>, JObj)
+              ,timestamp = wh_json:get_value(<<"timestamp">>, JObj)
+              ,ref_timestamp = wh_util:to_float(wh_json:get_value(<<"ref_timestamp">>, JObj))
+              ,label = wh_json:get_value(<<"label">>, JObj)
+              ,data = wh_json:get_value(<<"raw">>, JObj)
+              ,parser = wh_json:get_value(<<"parser">>, JObj)
+             }.
+
+-spec src(chunk()) -> {ne_binary(), pos_integer()}.
```
Comment (jamesaimonetti) [inline applications/call_inspector/src/parsers/ci_chunk.erl]: Takes a binary as well; same with dst/1

**Comment by fenollp** [comment]
@jamesaimonetti this is ready to merge

File: applications/call_inspector/src/parsers/ci_chunk.erl [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+
+-spec to_json(chunk()) -> wh_json:object().
+to_json(Chunk) ->
+    wh_json:from_list(
+      [{<<"call-id">>, call_id(Chunk)}
+       ,{<<"timestamp">>, timestamp(Chunk)}
+       ,{<<"ref_timestamp">>, wh_util:to_binary(ref_timestamp(Chunk))}
+       ,{<<"label">>, label(Chunk)}
+       ,{<<"raw">>, data(Chunk)}
+       ,{<<"src">>, src(Chunk)}
```
Comment (jamesaimonetti) [inline applications/call_inspector/src/parsers/ci_chunk.erl]: `src/1/` and `dst/1` appear to return 2-tuples. `{<<"src">>, {Host, Port}}` is not a valid JSON term. Should either be converted to a JSON object, combined into a binary, or something that is valid.

File: applications/call_inspector/src/ci_datastore.erl [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+    'true' = ci_analysis:is_analysis(Analysis),
+    CallId = ci_analysis:call_id(Analysis),
+    gen_server:cast(?SERVER, {'store_analysis', CallId, Analysis}).
+
+-spec callid_exists(ne_binary()) -> boolean().
+callid_exists(CallId) ->
+    File = make_name(CallId),
+    filelib:is_file(File).
+
+-type datum() :: {'chunk', [ci_chunk:chunk()]} |
```
Comment (jamesaimonetti) [inline applications/call_inspector/src/ci_datastore.erl]: You have `'chunk`' here; but in `ci_lookup_req` you access `'chunks'` from the `Props`. Which is it??? Make up your mind already!!! :)

**Comment by jamesaimonetti** [comment]
@fenollp Added some changes, spec updates, etc. Only things missing are the docs: `call_inspector/doc/index.md` with some basic setup help, and `crossbar/doc/call_inspector.md` for how to use the API. Everything else is looking good (though once you add setup docs I can actually test it!).

**Comment by fenollp** [comment]
@jamesaimonetti done at #1141

**Pull request merged**